### PR TITLE
fix(api): do not loose logs when api restart

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/cache/redis.go [hunk 1]
```diff
@@ -203,13 +227,13 @@ func (s *RedisStore) QueueLen(queueName string) (int, error) {
 }
 
 // DequeueWithContext gets from queue This is blocking while there is nothing in the queue, it can be cancelled with a context.Context
-func (s *RedisStore) DequeueWithContext(c context.Context, queueName string, value interface{}) error {
+func (s *RedisStore) DequeueWithContext(c context.Context, queueName string, waitDuration int64, value interface{}) error {
```
Comment (richardlt) [inline engine/api/cache/redis.go:230]: Can take a time.Duration

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13781.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/cdn/cdn_log.go [hunk 2]
```diff
@@ -324,3 +272,135 @@ func (s *Service) getHatchery(ctx context.Context, hatcheryID int64, hatcheryNam
...
+				return nil
+			}
+			// heartbeat
+			heartbeatKey := cache.Key(keyJobHearbeat, jobID)
+			if err := s.Cache.SetWithTTL(heartbeatKey, true, 30); err != nil {
+				log.Error(ctx, "unable to hearbeat %s: %v", heartbeatKey, err)
+				continue
+			}
+		default:
+			dequeuCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
```
Comment (richardlt) [inline engine/cdn/cdn_log.go:349]: What if for some reasons the heartbeat is not send or to slow, looks like another process will be able to dequeue but this one will not know about it.

Comment (sguiheux) [reply engine/cdn/cdn_log.go:349]: yes, he have 6 chances to heartbeat before another cdn take the queue

File: engine/cdn/cdn_log.go [hunk 3]
```diff
@@ -324,3 +272,135 @@ func (s *Service) getHatchery(ctx context.Context, hatcheryID int64, hatcheryNam
 	logCache.Set(fmt.Sprintf("hatchery-key-%d", hatcheryID), pk, gocache.DefaultExpiration)
 	return pk, nil
 }
+
+func (s *Service) waitingJobs(ctx context.Context) {
+	for {
```
Comment (fsamin) [inline engine/cdn/cdn_log.go:277]: we need a titicker

File: engine/cdn/cdn_log.go (hunk 2 above)
Comment (fsamin) [inline engine/cdn/cdn_log.go:349]: the cancel func must be called also in a defer

**Pull request merged**